File: frontend/src/types/knowledge.ts
export interface KnowledgeBaseItem {
  id: number
  department_id: number
  department_name: string
  code: string
  name: string
  description?: string | null
  visibility_scope: string
  status: string
  document_count: number
  created_by: number
  created_at: string
}

export interface KnowledgeBaseListResponse {
  items: KnowledgeBaseItem[]
  count: number
}

export interface CreateKnowledgeBasePayload {
  department_id: number
  code: string
  name: string
  description?: string
  visibility_scope: string
}

export interface KnowledgeDocumentItem {
  id: number
  knowledge_base_id: number
  knowledge_base_name: string
  department_id: number
  department_name: string
  title: string
  filename: string
  file_extension: string
  visibility_scope: string
  status: string
  chunks_count: number
  uploader_user_id: number
  uploader_username: string
  created_at: string
}

export interface KnowledgeDocumentListResponse {
  items: KnowledgeDocumentItem[]
  count: number
}

export interface UploadKnowledgeDocumentPayload {
  knowledge_base_id: number
  visibility_scope: string
  file: File
}

export interface UploadKnowledgeDocumentResponse {
  status: string
  message: string
  document: KnowledgeDocumentItem
}

export interface KnowledgeBaseShareItem {
  id: number
  target_department_id: number
  target_department_name: string
  status: string
  created_at: string
}

export interface KnowledgeBaseShareListResponse {
  items: KnowledgeBaseShareItem[]
  count: number
}

export interface CreateKnowledgeBaseShareRequestPayload {
  target_department_id: number
  reason: string
}

export interface KnowledgeBaseShareRequestItem {
  id: number
  knowledge_base_id: number
  knowledge_base_name: string
  source_department_id: number
  source_department_name: string
  target_department_id: number
  target_department_name: string
  requested_by: number
  requested_by_username: string
  requested_by_email: string
  reason: string
  status: string
  review_comment?: string | null
  submitted_at: string
  reviewed_at?: string | null
}

export interface KnowledgeBaseShareRequestListResponse {
  items: KnowledgeBaseShareRequestItem[]
  count: number
}

export interface ShareDecisionPayload {
  review_comment?: string
}

File: backend/app/models/knowledge.py
from sqlalchemy import Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.sql import func

from app.core.database import Base


class KnowledgeBase(Base):
    __tablename__ = "knowledge_bases"

    id = Column(Integer, primary_key=True, index=True)
    department_id = Column(Integer, ForeignKey("departments.id"), nullable=False, index=True)
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    code = Column(String, unique=True, index=True, nullable=False)
    name = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    visibility_scope = Column(String, default="department_private", nullable=False)
    status = Column(String, default="active", nullable=False, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())


class KnowledgeDocument(Base):
    __tablename__ = "knowledge_documents"

    id = Column(Integer, primary_key=True, index=True)
    knowledge_base_id = Column(Integer, ForeignKey("knowledge_bases.id"), nullable=False, index=True)
    department_id = Column(Integer, ForeignKey("departments.id"), nullable=False, index=True)
    uploader_user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    title = Column(String, nullable=False)
    filename = Column(String, nullable=False)
    file_extension = Column(String, nullable=False)
    visibility_scope = Column(String, default="knowledge_base_default", nullable=False)
    status = Column(String, default="active", nullable=False, index=True)
    chunks_count = Column(Integer, default=0, nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())


class KnowledgeBaseShare(Base):
    __tablename__ = "knowledge_base_shares"

    id = Column(Integer, primary_key=True, index=True)
    knowledge_base_id = Column(Integer, ForeignKey("knowledge_bases.id"), nullable=False, index=True)
    source_department_id = Column(Integer, ForeignKey("departments.id"), nullable=False, index=True)
    target_department_id = Column(Integer, ForeignKey("departments.id"), nullable=False, index=True)
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    status = Column(String, default="active", nullable=False, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())


class KnowledgeBaseShareRequest(Base):
    __tablename__ = "knowledge_base_share_requests"

    id = Column(Integer, primary_key=True, index=True)
    knowledge_base_id = Column(Integer, ForeignKey("knowledge_bases.id"), nullable=False, index=True)
    source_department_id = Column(Integer, ForeignKey("departments.id"), nullable=False, index=True)
    target_department_id = Column(Integer, ForeignKey("departments.id"), nullable=False, index=True)
    requested_by = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    reviewer_user_id = Column(Integer, ForeignKey("users.id"), nullable=True, index=True)
    reason = Column(Text, nullable=False)
    status = Column(String, default="pending", nullable=False, index=True)
    review_comment = Column(Text, nullable=True)
    submitted_at = Column(DateTime(timezone=True), server_default=func.now())
    reviewed_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
